sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/BufferingStreamObserver.java

```diff
@@ -105,10 +105,10 @@ public void onError(Throwable t) {
         // We check to see if we were able to successfully insert the poison pill at the front of
         // the queue to cancel the processing thread eagerly or if the processing thread is done.
         try {
-          // The order of these checks is important because short circuiting will cause us to
-          // insert into the queue first and only if it fails do we check that the thread is done.
-          while (!queue.offerFirst((T) POISON_PILL, 60, TimeUnit.SECONDS)
-              || !queueDrainer.isDone()) {
+          // We shouldn't attempt to insert into the queue if the queue drainer thread is done
+          // since the queue may be full and nothing will be emptying it.
+          while (!queueDrainer.isDone()
+              && !queue.offerFirst((T) POISON_PILL, 60, TimeUnit.SECONDS)) {
```

Comment:
Seems like there could be a tiny race here if the queue becomes done after checking but before offering, right?

Reply:
Nevermind, we'd just loop around again.

Reply:
Your correct, in the worst case the queue drainer thread will exit while the queue is full forcing us to wait the 60 seconds before checking again and then exiting. In the common case either the queue drainer is already done or the queue has room.

A better implementation would be to have a queue which can be told to finish itself without needing a poison pill but that is a change for another time.